File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.py
import sys
input = sys.stdin.readline

# 부모 노드 찾기
def find(x):
    if x !=parent[x]:
        return find(parent[x])
    return x

# 트리 합치기
def union(a, b):
    a = find(a)
    b = find(b)

    # 작은 루트 노드를 기준으로 합침
    if b < a:
        parent[a] = b
    else:
        parent[b] = a


v, e = map(int, input().split())

graph = [list(map(int, input().split())) for _ in range(e)]
# kruskal 알고리즘
graph.sort(key=lambda x: x[2]) # 간선들을 가중치 기준으로 정렬시킨다.

parent = [i for i in range(v+1)]
answer = 0

for s, e, w in graph:
    # 부모 노드가 같지 않다면 스패닝 트리
    if find(s) != find(e):
        union(s, e)
        answer += w

print(answer)

File: 백준/Gold/1197. 최소 스패닝 트리/최소 스패닝 트리.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {
    static int[] parent;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        int v = Integer.parseInt(st.nextToken());
        int e = Integer.parseInt(st.nextToken());

        parent = new int[v + 1];
        for (int i = 1; i <= v; i++) {
            parent[i] = i;
        }

        int[][] graph = new int[e][3];
        for (int i = 0; i < e; i++) {
            st = new StringTokenizer(br.readLine());
            graph[i][0] = Integer.parseInt(st.nextToken());
            graph[i][1] = Integer.parseInt(st.nextToken());
            graph[i][2] = Integer.parseInt(st.nextToken());
        }

        Arrays.sort(graph, (a, b) -> Integer.compare(a[2], b[2]));

        int ans = 0;
        for (int[] edge : graph) {
            int v1 = edge[0];
            int v2 = edge[1];
            int weight = edge[2];

            if (find(v1) != find(v2)) {
                union(v1, v2);
                ans += weight;
            }
        }

        System.out.println(ans);
    }

    static int find(int x) {
        if (x != parent[x]) {
            return find(parent[x]);
        }
        return x;
    }

    static void union(int a, int b) {
        a = find(a);
        b = find(b);

        if (a < b) {
            parent[b] = a;
        } else {
            parent[a] = b;
        }
    }
}
